R/get_session_script_code.R

```diff
@@ -0,0 +1,108 @@
+#' Generates a session script to replicate the App outputs
+#'
+#' @param template_path Path to the R script template (e.g., script_template.R)
+#' @param session The session object containing userData, etc.
+#' @param output_path Path to write the resulting script file (e.g., "output_script.R")
+#' @return The output_path (invisibly)
+get_session_script_code <- function(template_path, session, output_path) {
+  # Helper to get value from session$userData by path (e.g., 'settings$method')
+  get_session_value <- function(path) {
+    parts <- strsplit(path, "\\$")[[1]]
+    obj <- session$userData
+    for (p in parts) {
+      if (inherits(obj[[p]], "reactive")) {
+        obj <- obj[[p]]()
+      } else {
+        obj <- obj[[p]]
+      }
+      if (is.null(obj)) return(NULL)
+    }
+    obj
+  }
+
+  # Read template
+  script <- readLines(template_path, warn = FALSE) |>
+    paste(collapse = "\n")
+
+  # Find all session$userData$... or session$userData[[...]] or session$userData$...$...
+  # Regex for session$userData$foo or session$userData$foo$bar or session$userData[["foo"]]
+  pattern <- "session\\$userData(\\$[a-zA-Z0-9_]+(\\(\\))?(\\$[a-zA-Z0-9_]+)*)"
+  matches <- gregexpr(pattern, script, perl = TRUE)[[1]]
+  
+  # Just write the original template if there are no substitutions needed
+  # with a warning message
+  if (matches[1] == -1) {
+    script_lines <- strsplit(script, "\n")[[1]]
+    writeLines(script_lines, output_path)
+    warning(
+      "Please, the script template should have session$userData calls in order to work. ",
+      "Did you accidentally modify the original template?"
+    )
+    return(invisible(output_path))
+  }
+
+  # Replace all matches with deparsed values in one pass
+  match_positions <- gregexpr(pattern, script, perl = TRUE)[[1]]
+  match_strings <- regmatches(script, list(match_positions))
+  replacements <- lapply(match_strings, function(matched) {
+    path <- sub("^session\\$userData\\$", "", matched)
+    value <- get_session_value(path)
+    clean_deparse(value)
+  })
+  regmatches(script, list(match_positions)) <- replacements
+  # Split back into lines
+  script_lines <- strsplit(script, "\n")[[1]]
+  writeLines(script_lines, output_path)
+  invisible(output_path)
+}
+
+# Helper to cleanly deparse an object (data.frame, list, etc.)
+clean_deparse <- function(obj, indent = 0) {
+  ind <- paste(rep("  ", indent), collapse = "")
+  if (is.data.frame(obj)) {
+    # Multi-line, indented, no trailing comma on last column
+    cols <- lapply(obj, function(col) {
+      d <- paste(deparse(col, width.cutoff = 500), collapse = "")
+      if (grepl("^c\\(", d)) d else paste0("c(", d, ")")
```

Comment:
style: IMHO it is a bit hard to follow, something like this is less concise but I think a bit more readable:
```suggestion
      d <- col |>
        deparse(width.cutoff = 500) |>
        paste(collapse = "")
        
      if (!grepl("^c\\(", d)) {
        d <- paste0("c(", d, ")")
      }
      
      d
```